# NOTEBOOK 02 - Variables para el Análisis de Datos
## Semana 05 | Fundamentos de Python
### Profesor: Ing. Andrés Mena Abarca

**Nombre de los estudiantes:** __________________________________________

---

## Objetivo general

Analizar la utilización de variables y su importancia en el desarrollo de
algoritmos para el análisis de datos.

## Resultados de aprendizaje

Al finalizar esta actividad, podremos:

1. Diferenciar datos de entrada, variables de proceso e indicadores de salida.
2. Seleccionar nombres y tipos de datos apropiados.
3. Explicar las dependencias entre las variables de un algoritmo.
4. Calcular e interpretar indicadores básicos de ventas.
5. Utilizar una lista con `sum()`, `len()`, `max()` y `min()`.
6. Determinar mediante una variable booleana si se cumple una meta.

> En esta semana no estudiaremos ciclos. La lista se utilizará únicamente para
> reunir varios datos relacionados.

---

## 0. Puente desde las semanas anteriores

| Semana | Aprendizaje |
|---|---|
| 1 | Analizamos entradas, procesos y salidas. |
| 2 | Representamos datos mediante variables. |
| 3 | Usamos variables para tomar decisiones. |
| 4 | Limpiamos, convertimos y validamos datos. |
| 5 | Analizaremos cómo las variables producen indicadores útiles. |

Una variable no es solamente una caja. Su nombre, tipo, valor y posición en el
algoritmo deben comunicar qué representa el dato y cómo se utiliza.

### Actividad 0.1 - ¿El código comunica su propósito?

Analice este fragmento antes de ejecutarlo:

```python
x = [125000, 98000, 147000]
a = sum(x)
b = a / 3
```

Respondan en pareja:

1. ¿Qué podría representar `x`?
2. ¿Qué problemas causan los nombres `a` y `b`?
3. Propongan nombres que permitan comprender el algoritmo sin explicaciones.
4. Clasifiquen cada variable como dato original o dato derivado.

In [1]:
# Reescriba el ejemplo con nombres descriptivos.

ventas_diarias = [60000, 125000, 98000, 47000]
total_ventas = sum(ventas_diarias)
promedio_diario = total_ventas/len(ventas_diarias)

print(f"Ventas: {*ventas_diarias ,}")
print(f"Total: {total_ventas:,.2f}")
print(f"Promedio: {promedio_diario:,.2f}")

Ventas: (60000, 125000, 98000, 47000)
Total: 330,000.00
Promedio: 82,500.00


---

## 1. Variables con significado

En un algoritmo de análisis de datos podemos describir cada variable desde
**dos perspectivas diferentes**:

### A. Según su origen

| Categoría | Función | Ejemplo |
|---|---|---|
| Dato original | Conserva información recibida | `ventas_diarias` |
| Constante de referencia | Representa un valor que no debe cambiar durante el análisis | `META_SEMANAL` |
| Variable derivada | Se calcula utilizando otros datos | `total_ventas` |

### B. Según su propósito

| Categoría | Función | Ejemplo |
|---|---|---|
| Indicador | Resume o interpreta los datos | `promedio_diario` |
| Variable lógica | Contesta una pregunta con `True` o `False` | `meta_cumplida` |
| Salida interpretada | Comunica una conclusión | `conclusion` |

Una variable puede pertenecer a una categoría de cada perspectiva. Por ejemplo,
`promedio_diario` es una **variable derivada por su origen** y un
**indicador por su propósito**.

Usaremos mayúsculas para constantes y `snake_case` para las demás variables.

In [2]:
# Ejemplo guiado: variables de un análisis comercial
ventas_diarias = [125000, 98000, 147000, 132000, 110000]
META_SEMANAL = 600000

total_ventas = sum(ventas_diarias)
cantidad_dias = len(ventas_diarias)
promedio_diario = total_ventas / cantidad_dias
meta_cumplida = total_ventas >= META_SEMANAL

print(f"Total de ventas: ₡{total_ventas:,.2f}")
print(f"Promedio diario: ₡{promedio_diario:,.2f}")
print(f"¿Se cumplió la meta?: {meta_cumplida}")

Total de ventas: ₡612,000.00
Promedio diario: ₡122,400.00
¿Se cumplió la meta?: True


### Actividad 1.1 - Auditoría de variables

Complete la tabla a partir del ejemplo anterior.

| Variable | Tipo | Origen | Propósito | Significado |
|---|---|---|---|---|
| `ventas_diarias` | Lista | Dato original | Conjuntos a analizar | Ventas registradas |
| `META_SEMANAL` | int | Constante de referencia | Criterio de comparacion | Resultado esperado |
| `total_ventas` | int | Variable derivada | Indicador | Sumas ventas |
| `cantidad_dias` | int | Variable derivable | Indicador de cantidad | Número de dias registrados |
| `promedio_diario` | float | Variable derivada | Indicador | Venta media por día |
| `meta_cumplida` | bool | Variable derivable | Variable lógica | Indica cumplimiento |

Después, utilice `type()` para comprobar el tipo de dato de cada variable.

In [3]:
# Compruebe el tipo de cada variable con type().

print(type(ventas_diarias))
print(type(META_SEMANAL))
print(type(total_ventas))
print(type(cantidad_dias))
print(type(promedio_diario))
print(type(meta_cumplida))


<class 'list'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'float'>
<class 'bool'>


---

## 2. Dependencias y trazabilidad

Una variable derivada depende de otras variables. El orden del algoritmo es
importante porque Python necesita conocer los datos antes de calcular un
resultado.

```text
ventas_diarias ----> total_ventas ----+
       |                              |
       +----> cantidad_dias ----------+----> promedio_diario
       |
       +----> total_ventas ----+
                               +----> meta_cumplida
META_SEMANAL ------------------+
```

El promedio depende de `total_ventas` y `cantidad_dias`. La decisión depende de
`total_ventas` y `META_SEMANAL`.

Si cambia una venta, deben recalcularse todos los indicadores que dependen de
ella.

### Actividad 2.1 - Predicción antes de ejecutar

La segunda venta cambiará de `98000` a `150000`. En este ejercicio la meta será
`640000`, de modo que el cambio también permita observar una decisión distinta.

Antes de modificar el código, respondan:

1. ¿Aumentará o disminuirá `total_ventas`?
2. ¿Cambiará `cantidad_dias`?
3. ¿Cambiará `promedio_diario`?
4. ¿Cómo cambiará `meta_cumplida`?
5. ¿Cuáles variables deben recalcularse?

In [4]:
# Registre primero sus predicciones y luego compruébelas.
ventas_originales = [125000, 98000, 147000, 132000, 110000]
ventas_escenario = [125000, 150000, 147000, 132000, 110000]
META_ESCENARIO = 640000

# Complete los cálculos reemplazando cada espacio.
total_original = sum(ventas_originales)
promedio_original = total_original/len(ventas_originales)
meta_original = total_original >= META_ESCENARIO

total_escenario = sum(ventas_escenario)
promedio_escenario = total_escenario/len(ventas_escenario)
meta_escenario = total_escenario >= META_ESCENARIO


assert (
    total_original is not Ellipsis
    and promedio_original is not Ellipsis
    and meta_original is not Ellipsis
    and total_escenario is not Ellipsis
    and promedio_escenario is not Ellipsis
    and meta_escenario is not Ellipsis
), "Complete todos los cálculos del escenario antes de mostrar los resultados."

print(f"Total original: ₡{total_original:,.2f}")
print(f"Promedio original: ₡{promedio_original:,.2f}")
print(f"¿Meta original cumplida?: {meta_original}")
print(f"Total del escenario: ₡{total_escenario:,.2f}")
print(f"Promedio del escenario: ₡{promedio_escenario:,.2f}")
print(f"¿Meta del escenario cumplida?: {meta_escenario}")

Total original: ₡612,000.00
Promedio original: ₡122,400.00
¿Meta original cumplida?: False
Total del escenario: ₡664,000.00
Promedio del escenario: ₡132,800.00
¿Meta del escenario cumplida?: True


### Actividad 2.2 - Conservar el dato original

Observe estas dos opciones:

```python
# Opción A
ventas_brutas = 185000
devoluciones = 12000
ventas_brutas = ventas_brutas - devoluciones

# Opción B
ventas_brutas = 185000
devoluciones = 12000
ventas_netas = ventas_brutas - devoluciones
```

Respondan:

1. ¿Cuál opción conserva mejor la trazabilidad?
2. ¿Qué dato se pierde en la opción A?
3. ¿Cuál nombre explica mejor el resultado del cálculo?

In [ ]:
# Implemente la opción que conserva el dato original.
ventas_brutas = 185000
devoluciones = 12000

# ventas_netas = ...

# Muestre los tres valores con mensajes descriptivos.

---

## 3. De datos a indicadores

Una lista permite reunir valores relacionados:

```python
ventas_diarias = [125000, 98000, 147000, 132000, 110000]
```

Sin utilizar ciclos podemos obtener indicadores:

| Función | Resultado |
|---|---|
| `sum(datos)` | Suma de los valores |
| `len(datos)` | Cantidad de valores |
| `max(datos)` | Valor mayor |
| `min(datos)` | Valor menor |

El **rango** es la diferencia entre el valor mayor y el menor. Un rango elevado
indica mayor variación entre los datos.

### Actividad 3.1 - Construcción guiada del reporte

Complete los indicadores. Antes de programar, clasifique cada variable como
dato original, variable derivada o indicador.

In [ ]:
ventas_diarias = [125000, 98000, 147000, 132000, 110000]

# Complete las fórmulas reemplazando los puntos suspensivos.
# La celda producirá un error hasta que todas las fórmulas estén correctas.
total_ventas = sum(ventas_diarias)
cantidad_dias = len(ventas_diarias)
promedio_diario = total_ventas / cantidad_dias
venta_mayor = max(ventas_diarias)
venta_menor = min(ventas_diarias)
rango_ventas = venta_mayor - venta_menor


assert (
    total_ventas is not Ellipsis
    and cantidad_dias is not Ellipsis
    and promedio_diario is not Ellipsis
    and venta_mayor is not Ellipsis
    and venta_menor is not Ellipsis
    and rango_ventas is not Ellipsis
), "Complete todas las fórmulas antes de ejecutar el reporte."

print(f"Total: ₡{total_ventas:,.2f}")
print(f"Promedio: ₡{promedio_diario:,.2f}")
print(f"Venta mayor: ₡{venta_mayor:,.2f}")
print(f"Venta menor: ₡{venta_menor:,.2f}")
print(f"Rango: ₡{rango_ventas:,.2f}")

### Actividad 3.2 - Interpretación

Respondan con oraciones completas:

1. ¿Qué información aporta el promedio que no aporta el total?
2. ¿Qué significa la venta mayor para la empresa?
3. ¿Qué indica el rango sobre el comportamiento de las ventas?
4. ¿Un total alto garantiza que todos los días tuvieron ventas altas? Justifique.

> El rango utiliza únicamente el valor mayor y el menor. En esta actividad se
> emplea como una **primera aproximación a la variación**, no como una medida
> completa de estabilidad.

---

## 4. Detecte y corrija los errores

Analicen cada caso antes de ejecutar.

In [ ]:
# Error 1: 
d = [120000, 135000, 99000]
t = sum(d)

# Reescriba usando nombres descriptivos.

In [ ]:
# Error 2: 
venta_lunes = "125000"
venta_martes = 98000

# total_dos_dias = venta_lunes + venta_martes
# Explique el error.

In [ ]:
# Error 3: 
# Este ejemplo usa nombres nuevos para que no dependa de celdas anteriores.

# promedio_prueba = total_prueba / cantidad_prueba
# total_prueba = sum(datos_prueba)
# cantidad_prueba = len(datos_prueba)
# datos_prueba = [120000, 135000, 99000]

# Ordene y descomente las instrucciones para respetar las dependencias.

In [ ]:
# Error 4: sobrescritura que elimina trazabilidad
ventas_brutas = 200000
devoluciones = 15000
ventas_brutas = ventas_brutas - devoluciones

# Corrija el algoritmo conservando ventas_brutas.

---

## 5. Reto integrador en parejas: Diagnóstico semanal de ventas

Una pequeña empresa registró las ventas de lunes a viernes. La administración
necesita saber si alcanzó su meta semanal, cuánto le faltó o cuánto la superó
y qué información aportan las ventas diarias.

```python
ventas_semana = [145000, 118000, 162000, 137000, 128000]
META_SEMANAL = 700000
```

### Pregunta central

**¿La empresa alcanzó su meta semanal y qué indicadores permiten justificar
la conclusión?**

### Distribución del trabajo

**Participante A - Diseño del análisis**

- Propone los nombres y el orden de los cálculos.
- Selecciona y construye los indicadores que describen las ventas.
- Justifica cómo esos indicadores ayudan a responder la pregunta central.

**Participante B - Diseño de la decisión**

- Propone cómo comparar el resultado con la meta.
- Construye la decisión y la conclusión del reporte.
- Justifica qué información permite comunicar cuánto faltó o se superó.

**Trabajo conjunto:** integren ambos análisis, revisen las fórmulas del
compañero, prueben los escenarios y expliquen el algoritmo completo.

### Ruta de trabajo

Construyan un programa que:

1. Describa las ventas mediante cantidad, total, promedio, valor mayor,
   valor menor y rango.
2. Exprese qué porcentaje de la meta fue alcanzado.
3. Determine si la meta se cumplió y cuánto faltó o cuánto se superó.
4. Muestre un reporte claro y una conclusión comercial.
5. Conserve los datos originales y utilice nombres descriptivos.

La pareja decide los nombres de las variables, el orden de las instrucciones
y el formato del reporte. Antes de programar, acuerden el algoritmo en papel
o mediante un esquema breve.

### Criterios de evaluación

| Criterio | Peso |
|---|---:|
| Indicadores correctos | 30% |
| Nombres, tipos y clasificación | 20% |
| Decisión y diferencia respecto a la meta | 20% |
| Pruebas del algoritmo | 15% |
| Interpretación de los resultados | 15% |

In [10]:
# RETO INTEGRADOR
ventas_semana = [145000, 118000, 162000, 137000, 128000]
META_SEMANAL = 700000

# Construyan aquí el diagnóstico semanal.
# La solución debe responder la pregunta central y cumplir los requisitos.
cantidad_dias = len(ventas_semana)
total_ventas = sum(ventas_semana)
promedio_diario = total_ventas / cantidad_dias
venta_mayor = max(ventas_semana)
venta_menor = min(ventas_semana)
rango_ventas = venta_mayor - venta_menor

porcentaje_meta = (total_ventas / META_SEMANAL) * 100

diferencia_meta = total_ventas - META_SEMANAL

meta_superada = total_ventas > META_SEMANAL
meta_igualada = total_ventas == META_SEMANAL
meta_cumplida = total_ventas >= META_SEMANAL

assert (
    cantidad_dias is not Ellipsis
    and total_ventas is not Ellipsis
    and promedio_diario is not Ellipsis
    and venta_mayor is not Ellipsis
    and venta_menor is not Ellipsis
    and rango_ventas is not Ellipsis
    and porcentaje_meta is not Ellipsis
    and diferencia_meta is not Ellipsis
    and meta_cumplida is not Ellipsis
), "Complete todas las fórmulas antes de ejecutar el reporte."

print(f"Cantidad de días: {cantidad_dias}")
print(f"Total de ventas: ₡{total_ventas:,.2f}")
print(f"Promedio diario: ₡{promedio_diario:,.2f}")
print(f"Venta mayor: ₡{venta_mayor:,.2f}")
print(f"Venta menor: ₡{venta_menor:,.2f}")
print(f"Rango de ventas: ₡{rango_ventas:,.2f}")
print(f"Porcentaje de la meta alcanzado: {porcentaje_meta:.2f}%")
print(f"Diferencia respecto a la meta: ₡{diferencia_meta:,.2f}")

if meta_superada:
    print(f"Conclusión: la empresa SUPERÓ la meta semanal por ₡{diferencia_meta:,.2f}")
elif meta_igualada:
    print("Conclusión: la empresa IGUALÓ exactamente la meta semanal")
else:
    print(f"Conclusión: la empresa NO ALCANZÓ la meta semanal. Faltaron ₡{abs(diferencia_meta):,.2f}")


Cantidad de días: 5
Total de ventas: ₡690,000.00
Promedio diario: ₡138,000.00
Venta mayor: ₡162,000.00
Venta menor: ₡118,000.00
Rango de ventas: ₡44,000.00
Porcentaje de la meta alcanzado: 98.57%
Diferencia respecto a la meta: ₡-10,000.00
Conclusión: la empresa NO ALCANZÓ la meta semanal. Faltaron ₡10,000.00


### Registro e interpretación

Seleccionen **cinco variables creadas por la pareja** que sean importantes
para comprender el algoritmo. No incluyan `ventas_semana` ni `META_SEMANAL`.

| Variable elegida | Tipo | Origen | Propósito | Depende de |
|---|---|---|---|---|
| 1. total_ventas | int | Variable derivada | Conocer el monto total vendido en la semana | ventas_semana |
| 2. promedio_diario | float | Variable derivada | Conocer el comportamiento típico de ventas por día | total_ventas, cantidad_dias |
| 3. rango_ventas | int | Variable derivada | Medir qué tanto varían las ventas entre el mejor y el peor día | venta_mayor, venta_menor |
| 4. porcentaje_meta | float | Indicador | Expresar el avance hacia la meta de forma proporcional, no solo en colones | total_ventas, META_SEMANAL |
| 5. meta_cumplida | bool | Variable derivada (decisión) | Indicar si el total alcanzó o no la meta, para guiar la conclusión | total_ventas, META_SEMANAL |

Redacten una interpretación final de máximo tres oraciones que explique:

- si la meta se alcanzó y por qué;
- cuánto faltó o cuánto se superó;
- qué revela la variación de las ventas que el total no muestra.

Con los datos originales, la empresa no alcanzó la meta semanal, ya que vendió ₡690 000 frente a una meta de ₡700 000, quedando corta por ₡10 000 (un 98.57% de cumplimiento). El rango de ₡44 000 muestra que hubo variación moderada entre el día de mayores ventas (₡162 000) y el de menores ventas (₡118 000), lo cual revela que, aunque el total estuvo muy cerca de la meta, el desempeño no fue uniforme durante la semana.

### Pruebas obligatorias

Ejecuten el algoritmo con estos escenarios:

| Caso | Datos |
|---|---|
| Supera la meta | `[150000, 150000, 150000, 150000, 150000]` |
| Iguala la meta | `[140000, 140000, 140000, 140000, 140000]` |
| No alcanza la meta | `[100000, 110000, 120000, 130000, 140000]` |
| Alta variación | `[50000, 200000, 60000, 210000, 180000]` |

Antes de ejecutar cada caso, predigan la decisión y describan qué esperan
observar. Después, registren total, porcentaje, diferencia, rango y decisión.
Si el resultado no coincide con la predicción, expliquen la causa.

In [9]:
# PRUEBAS DEL ALGORITMO

# Caso 1: Supera la meta
ventas_caso1 = [150000, 150000, 150000, 150000, 150000]
total_caso1 = sum(ventas_caso1)
porcentaje_caso1 = (total_caso1 / META_SEMANAL) * 100
diferencia_caso1 = total_caso1 - META_SEMANAL
rango_caso1 = max(ventas_caso1) - min(ventas_caso1)

# Caso 2: Iguala la meta
ventas_caso2 = [140000, 140000, 140000, 140000, 140000]
total_caso2 = sum(ventas_caso2)
porcentaje_caso2 = (total_caso2 / META_SEMANAL) * 100
diferencia_caso2 = total_caso2 - META_SEMANAL
rango_caso2 = max(ventas_caso2) - min(ventas_caso2)

# Caso 3: No alcanza la meta
ventas_caso3 = [100000, 110000, 120000, 130000, 140000]
total_caso3 = sum(ventas_caso3)
porcentaje_caso3 = (total_caso3 / META_SEMANAL) * 100
diferencia_caso3 = total_caso3 - META_SEMANAL
rango_caso3 = max(ventas_caso3) - min(ventas_caso3)

# Caso 4: Alta variación
ventas_caso4 = [50000, 200000, 60000, 210000, 180000]
total_caso4 = sum(ventas_caso4)
porcentaje_caso4 = (total_caso4 / META_SEMANAL) * 100
diferencia_caso4 = total_caso4 - META_SEMANAL
rango_caso4 = max(ventas_caso4) - min(ventas_caso4)

assert (
    total_caso1 is not Ellipsis
    and total_caso2 is not Ellipsis
    and total_caso3 is not Ellipsis
    and total_caso4 is not Ellipsis
), "Complete todos los cálculos de los cuatro escenarios antes de mostrar los resultados."

print("Caso 1 - Supera la meta")
print(f"  Total: ₡{total_caso1:,.2f} | Porcentaje: {porcentaje_caso1:.2f}% | Diferencia: ₡{diferencia_caso1:,.2f} | Rango: ₡{rango_caso1:,.2f}")

print("Caso 2 - Iguala la meta")
print(f"  Total: ₡{total_caso2:,.2f} | Porcentaje: {porcentaje_caso2:.2f}% | Diferencia: ₡{diferencia_caso2:,.2f} | Rango: ₡{rango_caso2:,.2f}")

print("Caso 3 - No alcanza la meta")
print(f"  Total: ₡{total_caso3:,.2f} | Porcentaje: {porcentaje_caso3:.2f}% | Diferencia: ₡{diferencia_caso3:,.2f} | Rango: ₡{rango_caso3:,.2f}")

print("Caso 4 - Alta variación")
print(f"  Total: ₡{total_caso4:,.2f} | Porcentaje: {porcentaje_caso4:.2f}% | Diferencia: ₡{diferencia_caso4:,.2f} | Rango: ₡{rango_caso4:,.2f}")

Caso 1 - Supera la meta
  Total: ₡750,000.00 | Porcentaje: 107.14% | Diferencia: ₡50,000.00 | Rango: ₡0.00
Caso 2 - Iguala la meta
  Total: ₡700,000.00 | Porcentaje: 100.00% | Diferencia: ₡0.00 | Rango: ₡0.00
Caso 3 - No alcanza la meta
  Total: ₡600,000.00 | Porcentaje: 85.71% | Diferencia: ₡-100,000.00 | Rango: ₡40,000.00
Caso 4 - Alta variación
  Total: ₡700,000.00 | Porcentaje: 100.00% | Diferencia: ₡0.00 | Rango: ₡160,000.00


---

## 6. Cierre reflexivo

Respondan individualmente:

1. ¿Qué decisión de diseño tomaron sin recibir una plantilla y cómo la justifican?
   La decisión más importante que tomé fue distinguir entre "superar" la meta e "igualarla" exactamente, en lugar de usar solo una condición con >= que agrupara ambos casos como "meta cumplida". Aunque matemáticamente ambos casos satisfacen la meta, comercialmente no comunican lo mismo: superar la meta sugiere crecimiento, mientras que igualarla apenas sugiere estabilidad. Por eso usé meta_superada, meta_igualada y meta_cumplida como tres variables distintas, lo que me permitió escribir una conclusión más precisa con if/elif/else en lugar de un mensaje genérico de "sí o no".
2. ¿Qué indicador fue más útil para explicar el resultado y cuál fue su limitación?
   El indicador más útil fue el porcentaje de la meta alcanzado, porque traduce una cifra grande en colones a un número fácil de interpretar de inmediato: saber que se llegó al 98.57% comunica mucho más rápido qué tan cerca estuvo la empresa que ver el monto exacto en colones. Sin embargo, su limitación quedó clara en el caso de "alta variación": tanto ese escenario como el de "iguala la meta" arrojaron el mismo porcentaje (100%), pero representan situaciones de negocio completamente distintas. 

3. ¿Qué cambiarían en el algoritmo después de observar los casos de prueba?
   Después de ver que el caso de "alta variación" y el de "iguala la meta" daban el mismo resultado en cuanto a cumplimiento de la meta, pero con rangos completamente diferentes (₡0 frente a ₡160 000), agregaría una clasificación adicional sobre la estabilidad de las ventas, no solo sobre si se cumplió o no la meta. Por ejemplo, podría incluir una variable que indique si el rango es "bajo", "moderado" o "alto" en relación con el promedio diario, para que el reporte final no solo diga si la meta se cumplió, sino también qué tan parejo fue el desempeño durante la semana

## Lista de verificación

- [ ] Usamos nombres descriptivos en `snake_case`.
- [ ] Distinguimos datos originales, constantes y variables derivadas.
- [ ] Conservamos los datos originales.
- [ ] El algoritmo produce todos los indicadores solicitados.
- [ ] La decisión distingue correctamente superar, igualar y no alcanzar.
- [ ] El reporte comunica porcentaje, diferencia y variación con claridad.
- [ ] Predijimos y probamos cuatro escenarios sin conocer los resultados.
- [ ] Justificamos cinco variables creadas por la pareja.
- [ ] Ambos integrantes pueden explicar el algoritmo.